src/platform/posix/posix_pollq_epoll.c

```diff
@@ -0,0 +1,372 @@
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+static void
+nni_posix_poll_thr(void *arg)
+{
+	nni_posix_pollq *  pq = arg;
+	struct epoll_event events[NNI_MAX_EPOLL_EVENTS];
+
+	nni_mtx_lock(&pq->mtx);
+
+	while (!pq->close) {
+		int i;
+		int nevents;
+
+		// block indefinitely, timers are handled separately
+		nni_mtx_unlock(&pq->mtx);
+		nevents =
+		    epoll_wait(pq->epfd, events, NNI_MAX_EPOLL_EVENTS, -1);
+		nni_mtx_lock(&pq->mtx);
+
+		if (nevents <= 0) {
+			continue;
+		}
+
+		// dispatch events
+		for (i = 0; i < nevents; ++i) {
+			const struct epoll_event *ev;
+			nni_posix_pollq_node *    node;
+
+			ev = &events[i];
+			// If the waker pipe was signaled, read from it.
+			if ((ev->data.ptr == pq) && (ev->events & POLLIN)) {
+				nni_plat_pipe_clear(pq->wakerfd);
+				continue;
+			}
+
+			node = (nni_posix_pollq_node *) ev->data.ptr;
+			if (node->pq == NULL) {
+				// node was removed while we were blocking
+				continue;
+			}
+
+			node->revents = ev->events &
+			    (EPOLLIN | EPOLLOUT | EPOLLERR | EPOLLHUP);
+
+			// mark events as cleared
+			node->events &= ~node->revents;
+
+			// Save the active node; we can notice this way
+			// when it is busy, and avoid freeing it until
+			// we are sure that it is not in use.
+			pq->active = node;
+
+			// Execute the callback with lock released
+			nni_mtx_unlock(&pq->mtx);
+			node->cb(node->data);
+			nni_mtx_lock(&pq->mtx);
+
+			// We finished with this node.  If something
+			// was blocked waiting for that, wake it up.
+			pq->active = NULL;
+			if (pq->wait == node) {
+				pq->wait = NULL;
+				nni_cv_wake(&pq->cv);
+			}
+		}
+	}
+
+	nni_mtx_unlock(&pq->mtx);
+}
+
+static void
+nni_posix_pollq_destroy(nni_posix_pollq *pq)
+{
+	if (pq->started) {
+		nni_mtx_lock(&pq->mtx);
+		pq->close   = true;
+		pq->started = false;
+		nni_plat_pipe_raise(pq->wakewfd);
+		nni_mtx_unlock(&pq->mtx);
+	}
+	nni_thr_fini(&pq->thr);
+
+	if (pq->wakewfd >= 0) {
+		nni_plat_pipe_close(pq->wakewfd, pq->wakerfd);
+		pq->wakewfd = pq->wakerfd = -1;
+	}
+
+	close(pq->epfd);
+	pq->epfd = -1;
+
+	nni_mtx_fini(&pq->mtx);
+}
+
+static int
+nni_posix_pollq_add_wake_pipe(nni_posix_pollq *pq)
+{
+	// add user event so we can wake ourself on exit
+	struct epoll_event ev;
+	int                rv;
+
+	if (nni_plat_pipe_open(&pq->wakewfd, &pq->wakerfd) != 0) {
+		return (nni_plat_errno(errno));
+	}
+
+	ev.events   = EPOLLIN;
+	ev.data.ptr = pq;
+
+	rv = epoll_ctl(pq->epfd, EPOLL_CTL_ADD, pq->wakerfd, &ev);
+	if (rv != 0) {
+		return (nni_plat_errno(errno));
+	}
+	return (0);
+}
+
+static int
+nni_posix_pollq_create(nni_posix_pollq *pq)
+{
+	int rv;
+
+	if ((pq->epfd = epoll_create1(0)) < 0) {
+		return (nni_plat_errno(errno));
+	}
+
+	pq->wakewfd = -1;
+	pq->wakerfd = -1;
+	pq->close   = false;
+
+	nni_mtx_init(&pq->mtx);
+	nni_cv_init(&pq->cv, &pq->mtx);
+
+	if (((rv = nni_thr_init(&pq->thr, nni_posix_poll_thr, pq)) != 0) ||
+	    ((rv = nni_posix_pollq_add_wake_pipe(pq)) != 0)) {
```

Comment:
Thinking about this, rather than using the clunky wakewfd, wakerfd, why not just go ahead and use eventfd() directly.  We know its here, because we're on Linux.  (We shouldn't use epoll() anywhere else.)

Reply:
👍 implemented in 7d8e166